UCP/UCT/API: another client-server connection establishment flow

evgeny-leksikov opened the pull request:
This PR adds new UCP client-server connection establishment flow. The main difference from origin is possibility to handle connection requests on listener created on one worker but create an endpoint ("accept") on other worker. Also added ucp_listener_reject function to avoid memory leaks and all related UCT and testing things.

Please not, this PR is created against topic branch (not master) to make possible overview of the feature from top to bottom.

evgeny-leksikov committed:
UCP/API: add another client-server connection establishment flow

New listener accept callback passes EP address as an argument.
This address should be used for an endpoint creation directly connected
to the corresponding endpoint on client side, otherwise, ucp_ep_reject
should be called to notify client side by calling error handling
callback.
This approach allows to create listener on one worker but "accept"
connection requests by creating EPs on other worker.

evgeny-leksikov committed:
UCP/EP: add support UCP_EP_PARAM_FIELD_EP_ADDR

evgeny-leksikov committed:
GTEST: testing of UCP_EP_PARAM_FIELD_EP_ADDR flow

evgeny-leksikov committed:
UCT/API: add uct_iface_accept and uct_iface_reject

 - add new "id" arg to uct_sockaddr_conn_request_callback_t

evgeny-leksikov committed:
UCP/API: add ucp_listener_reject

evgeny-leksikov committed:
GTEST: testing of ucp_listener_reject

evgeny-leksikov committed:
UCT/RDMACM: fix coverity issue

evgeny-leksikov committed:
UCP/EP: suppress coverity issue

evgeny-leksikov committed:
GTEST: fix coverity issue and compilation with armclang

evgeny-leksikov committed:
UCP/DOC: improve ucp_listener_reject description

evgeny-leksikov committed:
UCP/LISTENER: apply CR comments

evgeny-leksikov committed:
UCP/EP: adjust parameters for UCP_EP_PARAM_FIELD_EP_ADDR case

evgeny-leksikov committed:
GTEST/UCP: testing of user data after ep creation on server side

evgeny-leksikov committed:
UCP/LISTENER: restore backward compatibility if accept CB is not set

evgeny-leksikov committed:
UCP/DOC: improve ucp_ep_params_t description

evgeny-leksikov committed:
UCP/EP: move dest_ep_ptr from gen extention to proto->conn

evgeny-leksikov committed:
UCP/LISTENER: protect worker in ucp_listener_conn_request_callback

evgeny-leksikov committed:
UCP/UCT/UCS: add new status code UCS_ERR_REJECTED

And replace UCS_ERR_CANCELED with UCS_ERR_REJECTED in client-server
connection establishment flow

evgeny-leksikov committed:
UCP/LISTENER: improve MT protection in ucp_listener_conn_request_callback

 - Currently, the callback can be called only from async thread of
   RDMACM TL but UCP can't assume this.
 - Other code review fixes

evgeny-leksikov committed:
UCT/API: connect/accept improvements

evgeny-leksikov committed:
UCT/RDMACM: connect/accept improvements

evgeny-leksikov committed:
GTESTS/UCT: fix connect/accept usage

evgeny-leksikov committed:
UCP/API: client/server improvements

evgeny-leksikov committed:
UCP/LISTENER: client/server improvements

evgeny-leksikov committed:
GTEST/UCP: client/server fixes

evgeny-leksikov committed:
GTEST/UCT: fix race condition in test_uct_sockaddr

evgeny-leksikov committed:
UCP/EP: reduce memory consumption for HPC cases (apply CR)

unite ep_match and ep_match fields of ucp_ep_ext_gen_t

evgeny-leksikov committed:
UCP/API: deprecate ucp_listener_accept_handler_t and ucp_ep_modify_nb

evgeny-leksikov committed:
Merge remote-tracking branch 'upstream/topic/client_server' into ucp_client_server_ep_address

evgeny-leksikov committed:
UCP/STREAM: fix recursive lock of async ctx on worker destroy

yosefe commented on src/ucp/api/ucp.h:
```diff
@@ -173,13 +173,19 @@ enum ucp_worker_params_field {
+
+    /**< User's callback and argument for handling the incoming connection
+     *   request. */
+    UCP_LISTENER_PARAM_FIELD_ACCEPT_CONN_HANDLER = UCS_BIT(2)
```
maybe we could use shorter name:
UCP_LISTENER_PARAM_FIELD_CONN_HANDLER

yosefe commented on src/ucp/api/ucp.h:
```diff
@@ -1669,6 +1588,23 @@ ucs_status_ptr_t ucp_ep_modify_nb(ucp_ep_h ep, const ucp_ep_params_t *params);
+ * @param [in]  conn_request    Handle to the connection request to reject.
+ *
+ */
+void ucp_listener_reject(ucp_listener_h listener,
```
maybe need to return a status? for future use

yosefe commented on src/ucp/api/ucp_compat.h:
```diff
@@ -85,6 +97,42 @@ ucs_status_t ucp_ep_flush(ucp_ep_h ep);
+ *
+ * @note See the documentation of @ref ucp_ep_params_t for details, only some of
+ *       the parameters can be modified.
+ * @deprecated Use @ref ucp_listener_accept_conn_handler_t instead of @ref
```
i think deprecated comment should be after `@brief`

yosefe commented on src/ucp/api/ucp_def.h:
```diff
@@ -438,4 +471,93 @@ typedef enum ucp_wakeup_event_types {
+
+    /**
+     * Destination address; one of the following fields is mandatory for
+     * filling:
```
this comment is tri-plicated, maybe better write this up in the documentation of ucp_ep_create

yosefe commented on src/ucp/core/ucp_ep.c:
```diff
@@ -75,9 +75,13 @@ ucs_status_t ucp_ep_new(ucp_worker_h worker, const char *peer_name,
     ucp_ep_ext_gen(ep)->err_cb      = NULL;
+    UCS_STATIC_ASSERT(sizeof(ucp_ep_ext_gen(ep)->ep_match) >=
+                      sizeof(ucp_ep_ext_gen(ep)->listener));
+    memset(&ucp_ep_ext_gen(ep)->ep_match, 0,
```
why?

evgeny-leksikov replied:
we initializes all old fields above, just to have whole struct initialized, this is a good practice and not perf critical since it's not a fast path.

yosefe commented on src/ucp/core/ucp_ep.c:
```diff
@@ -378,6 +382,88 @@ static ucs_status_t ucp_ep_create_to_sock_addr(ucp_worker_h worker,
     return status;
 }
 
+ucs_status_t ucp_ep_create_accept(ucp_worker_h worker,
```
pls add comment to what this function does

yosefe commented on src/ucp/core/ucp_ep.c:
```diff
@@ -378,6 +382,88 @@ static ucs_status_t ucp_ep_create_to_sock_addr(ucp_worker_h worker,
+
+    status = ucp_address_unpack(priv_addr + 1, &remote_address);
+    if (status != UCS_OK) {
+        return status;
```
since this function already uses goto, can add "goto out" and add label "out"

yosefe commented on src/ucp/core/ucp_ep.c:
```diff
@@ -378,6 +382,88 @@ static ucs_status_t ucp_ep_create_to_sock_addr(ucp_worker_h worker,
+    }
+
+    ucp_ep_update_dest_ep_ptr(*ep_p, priv_addr->ep_ptr);
+free_address:
```
out_free_address:

yosefe commented on src/ucp/core/ucp_ep.c:
```diff
@@ -378,6 +382,88 @@ static ucs_status_t ucp_ep_create_to_sock_addr(ucp_worker_h worker,
+
+static ucs_status_t
+ucp_ep_create_api_conn_request(ucp_worker_h worker,
+                               const ucp_ep_params_t *params, ucp_ep_h *ep_p)
```
params could be on 1st line

evgeny-leksikov replied:
Then `const ucp_ep_params_t *params` on the next line doesn't match 80 symbols requirement

yosefe commented on src/ucp/core/ucp_ep.c:
```diff
@@ -378,6 +382,88 @@ static ucs_status_t ucp_ep_create_to_sock_addr(ucp_worker_h worker,
+    /* coverity[overrun-buffer-val] */
+    status = ucp_ep_create_accept(worker, &conn_request->priv_addr, ep_p);
+    if (status != UCS_OK) {
+        return status;
```
goto out

yosefe commented on src/ucp/core/ucp_ep.c:
```diff
@@ -378,6 +382,88 @@ static ucs_status_t ucp_ep_create_to_sock_addr(ucp_worker_h worker,
+        return status;
+    }
+
+    ep = *ep_p;
```
better pass &ep in line 435

yosefe commented on src/ucp/core/ucp_ep.c:
```diff
@@ -378,6 +382,88 @@ static ucs_status_t ucp_ep_create_to_sock_addr(ucp_worker_h worker,
+    }
+
+    ep = *ep_p;
+    status = ucp_ep_adjust_params(ep, params);
```
can we pass params to ucp_ep_create_accept instead? let's avoid using 'adjsut_params', since ep_modify is now deprecated

evgeny-leksikov replied:
`adjsut_params` was added before `ep_modify `, it's kind of `ep` reconfiguration mechanism.

yosefe commented on src/ucp/core/ucp_ep.c:
```diff
@@ -378,6 +382,88 @@ static ucs_status_t ucp_ep_create_to_sock_addr(ucp_worker_h worker,
+
+    uct_iface_accept(conn_request->listener->wiface.iface, conn_request->id);
+
+    if (!(ep->flags & UCP_EP_FLAG_LISTENER)) {
```
could swap if/else to remove `!` here

yosefe commented on src/ucp/core/ucp_ep.c:
```diff
@@ -481,20 +567,20 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker, const ucp_ep_params_t *params,
-    } else {
+    } else if (params->field_mask & UCP_EP_PARAM_FIELD_CONN_REQUEST) {
+        status = ucp_ep_create_api_conn_request(worker, params, &ep);
+        ucs_free(params->conn_request);
```
why is this function releasing conn_request? looks like the inner function, ucp_ep_create_api_conn_request(), should be doing that..

yosefe commented on src/ucp/core/ucp_ep.c:
```diff
@@ -481,20 +567,20 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker, const ucp_ep_params_t *params,
-    *ep_p = ep;
+    if (status == UCS_OK) {
+        ep->flags |= UCP_EP_FLAG_USED;
+        *ep_p = ep;
```
indent on column

yosefe commented on src/ucp/core/ucp_ep.h:
```diff
@@ -275,20 +275,21 @@ typedef struct ucp_ep {
-    } ep_match;
+        } ep_match;
+
+        ucp_listener_h            listener;      /* Listener that may be associated with ep */
```
can you pls look at #2829, i am adding another field in union with ep_match to track flush for SW RMA, in theory it should be ok but need to be careful..

evgeny-leksikov replied:
Thanks! will double check on rebasing

yosefe commented on src/ucp/core/ucp_ep.h:
```diff
@@ -275,20 +275,21 @@ typedef struct ucp_ep {
  */
 typedef struct {
+    void                          *user_data;    /* User data associated with ep */
+    ucs_list_link_t               ep_list;       /* List entry in worker's all eps list */
```
let's put dest_ep_ptr first, it's used in fast-path protocols

yosefe commented on src/ucp/core/ucp_ep.h:
```diff
@@ -303,6 +304,24 @@ typedef struct {
+
+
+typedef struct ucp_conn_request {
+    struct ucp_listener         *listener;
```
ucp_listener_h

yosefe commented on src/ucp/core/ucp_listener.c:
```diff
@@ -121,44 +109,77 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
-                                                       const void *conn_priv_data,
-                                                       size_t length)
+static ucs_status_t
+ucp_listener_ep_conn_create(ucp_listener_h listener, void *id,
```
why converting 'id' to 'void *'? let's keep it uct_conn_request_h

yosefe commented on src/ucp/core/ucp_ep.h:
```diff
@@ -303,6 +304,24 @@ typedef struct {
+
+typedef struct ucp_conn_request {
+    struct ucp_listener         *listener;
+    void                        *id;
```
uct_conn_request_h instead of void*

yosefe commented on src/ucp/core/ucp_ep.h:
```diff
@@ -303,6 +304,24 @@ typedef struct {
+    uint8_t                   is_full_addr;  /**< Whether the attached address is
+                                                  full or partial */
+    /* packed worker address follows */
+} UCS_S_PACKED ucp_wireup_sockaddr_priv_t;
```
let's rename this struct to ucp_wireup_client_data_t

yosefe commented on src/ucp/core/ucp_ep.h:
```diff
@@ -303,6 +304,24 @@ typedef struct {
+typedef struct ucp_conn_request {
+    struct ucp_listener         *listener;
+    void                        *id;
+    ucp_wireup_sockaddr_priv_t  priv_addr;
```
rename priv_addr->client_data

yosefe commented on src/ucp/core/ucp_listener.h:
```diff
@@ -16,20 +16,36 @@
-    void                    *wireup_data;   /* Wireup private data*/
+    ucp_listener_h          listener;       /* Listener on which the connection
+                                               was accepted */
+    int                     is_ep;          /* flag to indicate which field is
```
i think union is not needed here, this structure size is not critical, can just have both ep and con_request and check which one is not NULL

yosefe commented on src/ucp/core/ucp_listener.c:
```diff
@@ -121,44 +109,77 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
+        accept->is_ep    = 1;
+        status = ucp_ep_create_accept(worker, client_data, &accept->ep);
+        if (status == UCS_OK) {
+            uct_iface_accept(tl_iface, conn_request);
```
pls add debug message to both cases

yosefe commented on src/ucp/core/ucp_listener.c:
```diff
@@ -121,44 +109,77 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
+    UCS_ASYNC_UNBLOCK(&worker->async);
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&worker->mt_lock);
+
+    ucs_assertv_always(status == UCS_OK, "connection request can't be handled");
```
weird, if we assert here why bother to reject?

yosefe commented on src/ucp/core/ucp_listener.c:
```diff
@@ -121,44 +109,77 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
+    accept->is_ep        = 0;
+    accept->conn_request = ucs_malloc(ucs_offsetof(ucp_conn_request_t,
+                                                   priv_addr) +
+                                      length, "accept connection request");
```
length should be on prev line

yosefe commented on src/ucp/core/ucp_listener.c:
```diff
@@ -121,44 +109,77 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
-        ucs_error("failed to allocate listener accept context");
-        return UCS_ERR_NO_MEMORY;
-    }
+    accept = ucs_malloc(sizeof(*accept), "ucp_listener_accept");
```
looks like we allocate both 'accept' and 'conn_request', can 'conn_request' be inside 'accept' and the use container_of, or some other way to allocate only once and not twice?

evgeny-leksikov replied:
No, because `ucp_conn_request_t` is not completed type. See
https://github.com/evgeny-leksikov/ucx/pull/3/files#diff-328d160be0fceb7e1c3a7d1f022c2ef3R321

yosefe commented on src/ucp/core/ucp_listener.c:
```diff
@@ -206,11 +233,22 @@ ucs_status_t ucp_listener_create(ucp_worker_h worker,
+            listener->arg       = params->accept_conn_handler.arg;
         } else {
-            listener->cb  = NULL;
+            listener->accept_cb = NULL;
```
maybe init all fields to NULL and then fill the rest inside each case?

yosefe commented on src/ucp/core/ucp_worker.c:
```diff
@@ -1190,13 +1190,9 @@ ucs_status_t ucp_worker_create(ucp_context_h context,
-    }
+    UCS_STATIC_ASSERT(sizeof(ucp_ep_ext_gen_t)   <= sizeof(ucp_ep_t));
+    UCS_STATIC_ASSERT(sizeof(ucp_ep_ext_proto_t) <= sizeof(ucp_ep_t));
+    ucs_strided_alloc_init(&worker->ep_alloc, sizeof(ucp_ep_t), 3);
```
we allocate 3 extensions even for HPC??

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1266,6 +1266,31 @@ ucs_status_t uct_iface_set_am_tracer(uct_iface_h iface, uct_am_tracer_t tracer,
+ * @param [in] conn_request Connection establishment request passed as parameter
+ *                          of @ref uct_sockaddr_conn_request_callback_t.
+ */
+void uct_iface_accept(uct_iface_h iface, uct_conn_request_h conn_request);
```
let's return status code from these functions

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1266,6 +1266,31 @@ ucs_status_t uct_iface_set_am_tracer(uct_iface_h iface, uct_am_tracer_t tracer,
+ * @param [in] conn_request Connection establishment request passed as parameter
+ *                          of @ref uct_sockaddr_conn_request_callback_t.
+ */
+void uct_iface_reject(uct_iface_h iface, uct_conn_request_h conn_request);
```
return status code

yosefe commented on src/uct/ib/rdmacm/rdmacm_iface.c:
```diff
@@ -56,12 +56,54 @@ static ucs_status_t uct_rdmacm_iface_get_address(uct_iface_h tl_iface, uct_iface
     return UCS_OK;
 }
 
+void uct_rdmacm_accept(void *id)
```
rdmacm_id instead of void*?

yosefe commented on src/uct/ib/rdmacm/rdmacm_iface.c:
```diff
@@ -56,12 +56,54 @@ static ucs_status_t uct_rdmacm_iface_get_address(uct_iface_h tl_iface, uct_iface
+    /* Accepting the connection will generate the RDMA_CM_EVENT_ESTABLISHED
+     * event on the client side. */
+    if (rdma_accept(id, &conn_param)) {
+        ucs_error("rdma_accept(to id=%p) failed: %m.", id);
```
no need for '.' in the end

yosefe commented on src/uct/ib/rdmacm/rdmacm_iface.c:
```diff
@@ -56,12 +56,54 @@ static ucs_status_t uct_rdmacm_iface_get_address(uct_iface_h tl_iface, uct_iface
+     * event on the client side. */
+    if (rdma_accept(id, &conn_param)) {
+        ucs_error("rdma_accept(to id=%p) failed: %m.", id);
+        rdma_reject(id, NULL, 0);
```
do we really need rdma_reject if failed to accept?

yosefe commented on src/uct/ib/rdmacm/rdmacm_iface.c:
```diff
@@ -192,7 +213,14 @@ static void uct_rdmacm_iface_cm_id_to_dev_name(struct rdma_cm_id *cm_id,
 
-static int uct_rdmacm_iface_process_event(uct_rdmacm_iface_t *iface, struct rdma_cm_event *event)
+typedef struct uct_rdmacm_process_event_ret {
+    uint8_t destroy_cm_id;
```
can we use flags instead?

yosefe commented on src/uct/ib/rdmacm/rdmacm_iface.c:
```diff
@@ -203,7 +231,11 @@ static int uct_rdmacm_iface_process_event(uct_rdmacm_iface_t *iface, struct rdma
     uct_rdmacm_ep_t *ep = NULL;
     ssize_t priv_data_ret;
-    int destroy_cm_id = 0;
+    ucs_status_t status;
```
looks like 'status' variable is not really needed

evgeny-leksikov replied:
Why? It's needed to handle `RDMA_CM_EVENT_UNREACHABLE` when it should be `RDMA_CM_EVENT_REJECTED` but not supported for `RDMA_PS_UDP`

yosefe commented on src/uct/ib/rdmacm/rdmacm_iface.c:
```diff
@@ -281,39 +314,49 @@ static int uct_rdmacm_iface_process_event(uct_rdmacm_iface_t *iface, struct rdma
     /* client error events */
+    case RDMA_CM_EVENT_UNREACHABLE:
+        if ((event->param.conn.private_data_len > 0) &&
+            !strncmp("reject", (char *)event->param.conn.private_data,
```
huh?

evgeny-leksikov replied:
This is the only way for `RDMA_PS_UDP`

yosefe commented on test/gtest/ucp/test_ucp_sockaddr.cc:
```diff
@@ -23,29 +23,41 @@
-                                                        const std::string& test_case_name,
-                                                        const std::string& tls)
+    enum {
+        MT_PARAM_VARIANT = DEFAULT_PARAM_VARIANT + 1,
```
can u pls comment what each test mode means

yosefe commented on test/gtest/ucp/test_ucp_sockaddr.cc:
```diff
@@ -222,13 +259,37 @@ class test_ucp_sockaddr : public ucp_test {
+                                       1, 0, rtag_complete_cb);
         } else {
+            ASSERT_TRUE(send_recv_type == SEND_RECV_STREAM);
+            ucp_stream_poll_ep_t poll_eps[1];
```
it doesn't have to be an array..

yosefe commented on test/gtest/ucp/test_ucp_sockaddr.cc:
```diff
@@ -366,7 +511,15 @@ class test_ucp_sockaddr_with_wakeup : public test_ucp_sockaddr {
+
+UCS_TEST_P(test_ucp_sockaddr_with_wakeup, reject) {
+    if (GetParam().variant > 0) {
+        UCS_TEST_SKIP_R("Not parameterized test");
```
this message seems weird.. maybe you meant invalid test parameter?

yosefe commented on test/gtest/ucp/ucp_test.cc:
```diff
@@ -91,6 +92,21 @@ ucp_test_base::entity* ucp_test::create_entity(bool add_in_front,
+    for (e_it = entities().begin(); e_it != entities().end(); ++e_it) {
+        for (int w_idx = 0; w_idx < (*e_it)->get_num_workers(); ++w_idx) {
+            for (int ep_idx = 0; ep_idx < (*e_it)->get_num_eps(w_idx); ++ep_idx)
+            {
```
{ on prev line

yosefe commented on test/gtest/ucp/ucp_test.cc:
```diff
@@ -477,18 +523,31 @@ ucp_ep_h ucp_test_base::entity::revoke_ep(int worker_index, int ep_index) const
+    params.field_mask             = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR;
+    params.sockaddr.addr          = saddr;
+    params.sockaddr.addrlen       = addrlen;
+    if (cb_type == LISTEN_CB_EP) {
```
switch/case ?

yosefe commented on test/gtest/ucp/ucp_test.cc:
```diff
@@ -518,7 +577,21 @@ ucp_context_h ucp_test_base::entity::ucph() const {
+        return 0;
+    }
+
+    unsigned ret = 0;
```
rename ret->count/progress_count

yosefe commented on test/gtest/ucp/ucp_test.h:
```diff
@@ -46,6 +48,12 @@ class ucp_test_base : public ucs::test_base {
 
     public:
+        typedef enum {
+            LISTEN_CB_EP,
```
pls comment what those mean

yosefe commented on test/gtest/uct/ib/test_sockaddr.cc:
```diff
@@ -120,12 +133,57 @@ UCS_TEST_P(test_uct_sockaddr, connect_client_to_server)
+        uct_iface_accept(server->iface(), delayed_conn_reqs.front());
+        delayed_conn_reqs.pop();
+    }
+    sleep(3);
```
why sleep needed? can we have a condition check instead?

evgeny-leksikov replied:
Copy/pasted from https://github.com/evgeny-leksikov/ucx/pull/3/files#diff-93fd01f353a7644c48f98fd17ae69187R277
I guess condition check requires transport level hooks

yosefe commented:
@evgeny-leksikov added few comments

yosefe commented on src/ucp/core/ucp_listener.c:
```diff
@@ -121,44 +109,77 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    if (listener->conn_cb != NULL) {
+        status = ucp_listener_ep_conn_create(listener, conn_request,
```
merge issue - can deadlock - when taking worker lock

evgeny-leksikov committed:
UCT/UCP/TESTS: client-server connection establishment. Apply CR comments

evgeny-leksikov committed:
UCP/LISTENER: fix merge issue in accept flow

evgeny-leksikov committed:
UCP/GTESTS: remove extra out-of-band reachability check (TODO)

evgeny-leksikov committed:
Merge remote-tracking branch 'upstream/master' into ucp_client_server_ep_address

Conflicts:
	src/ucp/core/ucp_ep.h
	src/ucp/core/ucp_listener.c

evgeny-leksikov committed:
UCP/GTESTS: remove extra out-of-band reachability check

evgeny-leksikov committed:
UCP/WIREUP: improve err handling and bugfix

 - invoke err handling flow for client-server connection establishment
   if ucp_wireup_process_pre_request fails with UCS_ERR_UNREACHABLE on
   lanes initialization
 - fix memory corruption in ucp_ep_ext_gen_t::enum across flush_state
   and listener fields

evgeny-leksikov committed:
Merge remote-tracking branch 'upstream/master' into ucp_client_server_ep_address

Conflicts:
	src/ucp/api/ucp.h
	src/ucp/api/ucp_compat.h
	src/ucp/api/ucp_def.h
	test/gtest/ucp/test_ucp_sockaddr.cc
